File: mobile/native/darwin/Packages/EnteNetwork/Sources/EnteNetwork/Gateways/AuthenticationGateway.swift
import Foundation
import EnteCore
import Logging

// MARK: - Authentication Models

public struct SRPAttributes: Codable {
    public let srpUserID: String
    public let srpSalt: String
    public let kekSalt: String
    public let memLimit: Int
    public let opsLimit: Int
    public let isEmailMFAEnabled: Bool
    
    public init(srpUserID: String, srpSalt: String, kekSalt: String, memLimit: Int, opsLimit: Int, isEmailMFAEnabled: Bool) {
        self.srpUserID = srpUserID
        self.srpSalt = srpSalt
        self.kekSalt = kekSalt
        self.memLimit = memLimit
        self.opsLimit = opsLimit
        self.isEmailMFAEnabled = isEmailMFAEnabled
    }
}

public struct CreateSRPSessionResponse: Codable {
    public let sessionID: String
    public let srpB: String
    
    public init(sessionID: String, srpB: String) {
        self.sessionID = sessionID
        self.srpB = srpB
    }
}

public struct KeyAttributes: Codable {
    public let kekSalt: String
    public let encryptedKey: String
    public let keyDecryptionNonce: String
    public let publicKey: String
    public let encryptedSecretKey: String
    public let secretKeyDecryptionNonce: String
    public let memLimit: Int
    public let opsLimit: Int
    
    public init(kekSalt: String, encryptedKey: String, keyDecryptionNonce: String, publicKey: String, encryptedSecretKey: String, secretKeyDecryptionNonce: String, memLimit: Int, opsLimit: Int) {
        self.kekSalt = kekSalt
        self.encryptedKey = encryptedKey
        self.keyDecryptionNonce = keyDecryptionNonce
        self.publicKey = publicKey
        self.encryptedSecretKey = encryptedSecretKey
        self.secretKeyDecryptionNonce = secretKeyDecryptionNonce
        self.memLimit = memLimit
        self.opsLimit = opsLimit
    }
}

public struct AuthorizationResponse: Codable {
    public let keyAttributes: KeyAttributes?
    public let encryptedToken: String?
    public let token: String?
    public let twoFactorSessionID: String?
    public let twoFactorSessionIDV2: String?
    public let passkeySessionID: String?
    public let accountsUrl: String?
    public let id: UserID
    
    public init(keyAttributes: KeyAttributes?, encryptedToken: String?, token: String?, twoFactorSessionID: String?, twoFactorSessionIDV2: String?, passkeySessionID: String?, accountsUrl: String?, id: UserID) {
        self.keyAttributes = keyAttributes
        self.encryptedToken = encryptedToken
        self.token = token
        self.twoFactorSessionID = twoFactorSessionID
        self.twoFactorSessionIDV2 = twoFactorSessionIDV2
        self.passkeySessionID = passkeySessionID
        self.accountsUrl = accountsUrl
        self.id = id
    }
    
    public var effectiveTwoFactorSessionID: String? {
        return twoFactorSessionIDV2 ?? twoFactorSessionID
    }
    
    public var isMFARequired: Bool {
        return effectiveTwoFactorSessionID?.isEmpty == false
    }
    
    public var isPasskeyRequired: Bool {
        return passkeySessionID?.isEmpty == false
    }
}

// MARK: - Authentication Gateway

public class AuthenticationGateway {
    private let client: APIClient
    private let logger = Logger(label: "AuthenticationGateway")
    
    internal init(client: APIClient) {
        self.client = client
    }
    
    // MARK: - SRP Authentication
    
    public func getSRPAttributes(email: String) async throws -> SRPAttributes {
        logger.info("Getting SRP attributes for email")
        
        struct Response: Codable {
            let attributes: SRPAttributes
        }
        
        let response = try await client.request(
            AuthEndpoint.getSRPAttributes(email: email),
            responseType: Response.self
        )
        
        return response.attributes
    }
    
    public func createSRPSession(srpUserID: String, clientPub: String) async throws -> CreateSRPSessionResponse {
        logger.info("Creating SRP session")
        
        return try await client.request(
            AuthEndpoint.createSRPSession(srpUserID: srpUserID, clientPub: clientPub),
            responseType: CreateSRPSessionResponse.self
        )
    }
    
    public func verifySRPSession(srpUserID: String, sessionID: String, clientM1: String) async throws -> AuthorizationResponse {
        logger.info("Verifying SRP session")
        
        return try await client.request(
            AuthEndpoint.verifySRPSession(srpUserID: srpUserID, sessionID: sessionID, clientM1: clientM1),
            responseType: AuthorizationResponse.self
        )
    }
    
    // MARK: - Email OTP
    
    public func sendLoginOTP(email: String, purpose: String = "login") async throws {
        logger.info("Sending login OTP")
        
        try await client.request(AuthEndpoint.sendLoginOTP(email: email, purpose: purpose))
    }
    
    public func verifyEmail(email: String, otp: String) async throws -> AuthorizationResponse {
        logger.info("Verifying email with OTP")
        
        return try await client.request(
            AuthEndpoint.verifyEmail(email: email, otp: otp),
            responseType: AuthorizationResponse.self
        )
    }
    
    // MARK: - Two-Factor Authentication
    
    public func verifyTOTP(sessionID: String, otp: String) async throws -> AuthorizationResponse {
        logger.info("Verifying TOTP")
        
        return try await client.request(
            AuthEndpoint.verifyTOTP(sessionID: sessionID, otp: otp),
            responseType: AuthorizationResponse.self
        )
    }
    
    // MARK: - Passkeys
    
    public func getTokenForPasskeySession(sessionID: String) async throws -> AuthorizationResponse {
        logger.info("Getting token for passkey session")
        
        return try await client.request(
            AuthEndpoint.getTokenForPasskeySession(sessionID: sessionID),
            responseType: AuthorizationResponse.self
        )
    }
}

File: mobile/native/android/apps/ensu/data/src/main/java/io/ente/ensu/data/network/AuthenticationGateway.kt
package io.ente.ensu.data.network

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

internal class AuthenticationGateway(private val client: ApiClient) {
    suspend fun getSrpAttributes(email: String): SrpAttributesResponse {
        return client.request(AuthEndpoint.GetSrpAttributes(email), SrpAttributesResponse.serializer())
    }

    suspend fun createSrpSession(srpUserId: String, clientPub: String): CreateSrpSessionResponse {
        return client.request(
            AuthEndpoint.CreateSrpSession(srpUserId, clientPub),
            CreateSrpSessionResponse.serializer()
        )
    }

    suspend fun verifySrpSession(srpUserId: String, sessionId: String, clientM1: String): AuthorizationResponse {
        return client.request(
            AuthEndpoint.VerifySrpSession(srpUserId, sessionId, clientM1),
            AuthorizationResponse.serializer()
        )
    }

    suspend fun sendLoginOtp(email: String, purpose: String = "login") {
        client.request(AuthEndpoint.SendLoginOtp(email, purpose))
    }

    suspend fun verifyEmail(email: String, otp: String): AuthorizationResponse {
        return client.request(AuthEndpoint.VerifyEmail(email, otp), AuthorizationResponse.serializer())
    }

    suspend fun verifyTotp(sessionId: String, otp: String): AuthorizationResponse {
        return client.request(AuthEndpoint.VerifyTotp(sessionId, otp), AuthorizationResponse.serializer())
    }

    suspend fun getTokenForPasskeySession(sessionId: String): AuthorizationResponse {
        return client.request(
            AuthEndpoint.GetTokenForPasskeySession(sessionId),
            AuthorizationResponse.serializer()
        )
    }
}

@Serializable
internal data class SrpAttributesResponse(
    @SerialName("attributes") val attributes: SrpAttributesDto
)

@Serializable
internal data class SrpAttributesDto(
    @SerialName("srpUserID") val srpUserId: String,
    @SerialName("srpSalt") val srpSalt: String,
    @SerialName("kekSalt") val kekSalt: String,
    @SerialName("memLimit") val memLimit: Int,
    @SerialName("opsLimit") val opsLimit: Int,
    @SerialName("isEmailMFAEnabled") val isEmailMfaEnabled: Boolean
)

@Serializable
internal data class CreateSrpSessionResponse(
    @SerialName("sessionID") val sessionId: String,
    @SerialName("srpB") val srpB: String
)

@Serializable
internal data class KeyAttributesDto(
    @SerialName("kekSalt") val kekSalt: String,
    @SerialName("encryptedKey") val encryptedKey: String,
    @SerialName("keyDecryptionNonce") val keyDecryptionNonce: String,
    @SerialName("publicKey") val publicKey: String,
    @SerialName("encryptedSecretKey") val encryptedSecretKey: String,
    @SerialName("secretKeyDecryptionNonce") val secretKeyDecryptionNonce: String,
    @SerialName("memLimit") val memLimit: Int? = null,
    @SerialName("opsLimit") val opsLimit: Int? = null
)

@Serializable
internal data class AuthorizationResponse(
    @SerialName("keyAttributes") val keyAttributes: KeyAttributesDto? = null,
    @SerialName("encryptedToken") val encryptedToken: String? = null,
    @SerialName("token") val token: String? = null,
    @SerialName("twoFactorSessionID") val twoFactorSessionId: String? = null,
    @SerialName("twoFactorSessionIDV2") val twoFactorSessionIdV2: String? = null,
    @SerialName("passkeySessionID") val passkeySessionId: String? = null,
    @SerialName("accountsUrl") val accountsUrl: String? = null,
    @SerialName("id") val id: Long
) {
    val effectiveTwoFactorSessionId: String?
        get() = twoFactorSessionIdV2 ?: twoFactorSessionId
}
